### PR TITLE
mgr/volumes: validate subvolume create cmd args

### PR DESCRIPTION
**Pull request by mchangir**
test subvolgroup and subvol names and declare empty/whitespace strings as bad names

Fixes: https://tracker.ceph.com/issues/58645
Signed-off-by: Milind Changire <mchangir@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by mchangir** [comment]
Update:
Replaced word `group` with `subvolgroup` in PR description and commit message.

File: src/pybind/mgr/volumes/fs/operations/group.py [hunk 1]
```diff
@@ -22,6 +22,8 @@ class Group(GroupTemplate):
     def __init__(self, fs, vol_spec, groupname):
         if groupname == Group.NO_GROUP_NAME:
             raise VolumeException(-errno.EPERM, "Operation not permitted for group '{0}' as it is an internal group.".format(groupname))
+        if groupname and groupname.lstrip() == "":
+            raise VolumeException(-errno.EINVAL, f"'{groupname}' is a bad group name")
```
Comment (joscollin) [inline src/pybind/mgr/volumes/fs/operations/group.py:26]: just using `isspace()` is better?

Comment (mchangir) [reply src/pybind/mgr/volumes/fs/operations/group.py:26]: I need to test for **empty string** (**""**) passed in from command-line as well.

Comment (neesingh-rh) [reply src/pybind/mgr/volumes/fs/operations/group.py:26]: If that's the case, then I guess it should be `if not groupname....` otherwise for `empty string ( "" )` Exception won't be raised.

Comment (neesingh-rh) [reply src/pybind/mgr/volumes/fs/operations/group.py:26]: I guess we can have `if  not( groupname and groupname.strip()):`

Comment (neesingh-rh) [reply src/pybind/mgr/volumes/fs/operations/group.py:26]: > I need to test for **empty string** (**""**) passed in from command-line as well.

I referred to this comment. If u say we don't need to check for `empty string("")` then `isspace()` is better.

Comment (mchangir) [reply src/pybind/mgr/volumes/fs/operations/group.py:26]: > > I need to test for **empty string** (**""**) passed in from command-line as well.
> 
> I referred to this comment. If u say we don't need to check for `empty string("")` then `isspace()` is better.

As I said earlier, I **do** need to test for empty string as well.
And `isspace()` returns `False` for an empty string.

**Comment by vshankar** [comment]
@mchangir Do we know why the string (subvolume, group) isn't caught before even getting called into the volumes driver?

**Comment by mchangir** [comment]
> @mchangir Do we know why the string (subvolume, group) isn't caught before even getting called into the volumes driver

AFAICT, its the volumes driver that enforces all the rules on the subvolume and group naming, which, I think, is the right place do such validations

**Comment by rishabh-d-dave** [comment]
jenkins test make check

**Review by rishabh-d-dave — COMMENTED** [review]
Looks okay to me.

**Comment by vshankar** [comment]
> > @mchangir Do we know why the string (subvolume, group) isn't caught before even getting called into the volumes driver
> 
> AFAICT, its the volumes driver that enforces all the rules on the subvolume and group naming, which, I think, is the right place do such validations

Right. However, in this case its a empty string that getting passed in. I expected this to be caught before calling into the volumes driver.

**Comment by mchangir** [comment]
> > > @mchangir Do we know why the string (subvolume, group) isn't caught before even getting called into the volumes driver
> > 
> > 
> > AFAICT, its the volumes driver that enforces all the rules on the subvolume and group naming, which, I think, is the right place do such validations
> 
> Right. However, in this case its a empty string that getting passed in. I expected this to be caught before calling into the volumes driver.

There's a way to mention what are the acceptable characters for a parameter via the `goodchars` spec.

Is there a way to mention in the command parameter spec that the argument also needs to be non-empty string ?

Where can I find the parameter spec attributes definitions ?
Maybe I could add a spec to specify non-empty string.

**Comment by vshankar** [comment]
> > > > @mchangir Do we know why the string (subvolume, group) isn't caught before even getting called into the volumes driver
> > > 
> > > 
> > > AFAICT, its the volumes driver that enforces all the rules on the subvolume and group naming, which, I think, is the right place do such validations
> > 
> > 
> > Right. However, in this case its a empty string that getting passed in. I expected this to be caught before calling into the volumes driver.
> 
> There's a way to mention what are the acceptable characters for a parameter via the `goodchars` spec.
> 
> Is there a way to mention in the command parameter spec that the argument also needs to be non-empty string ?
> 
> Where can I find the parameter spec attributes definitions ? Maybe I could add a spec to specify non-empty string.

Something along those lines. I think other mgr modules just might have the similar issues with empty strings.

**Comment by mchangir** [comment]
> > > > > @mchangir Do we know why the string (subvolume, group) isn't caught before even getting called into the volumes driver
> > > > 
> > > > 
> > > > AFAICT, its the volumes driver that enforces all the rules on the subvolume and group naming, which, I think, is the right place do such validations
> > > 
> > > 
> > > Right. However, in this case its a empty string that getting passed in. I expected this to be caught before calling into the volumes driver.
> > 
> > 
> > There's a way to mention what are the acceptable characters for a parameter via the `goodchars` spec.
> > Is there a way to mention in the command parameter spec that the argument also needs to be non-empty string ?
> > Where can I find the parameter spec attributes definitions ? Maybe I could add a spec to specify non-empty string.
> 
> Something along those lines. I think other mgr modules just might have the similar issues with empty strings.

I looked up other `mgr/volumes` **create** commands. The only ones found were:
* fs subvolumegroup create
* fs subvolumegroup snapshot create
* fs subvolume create
* fs subvolume snapshot create

The `subvolume* snapshot create` commands fail with various errors without explicit argument validation and hence haven't had this issue before.

e.g. `fs snap-schedule add` command never validates input arguments for `fs` and `path`, the effect of validation only takes place when the actual snapshot creation is attempted and then the snap-schedule is deactivated or when the `INSERT` into the DB happens via the DB field/column constraints

**Comment by dparmar18** [comment]
changes in `src/pybind/ceph_argparse.py` seems to have broken something:

```
FAIL: test_parse_json_funcsigs (__main__.ParseJsonFuncsigs)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/pybind/test_ceph_argparse.py", line 46, in test_parse_json_funcsigs
    self.assertRaises(TypeError, parse_json_funcsigs, commands, 'cli')
AssertionError: TypeError not raised by parse_json_funcsigs

----------------------------------------------------------------------
Ran 136 tests in 58.276s

FAILED (failures=1)
sys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name=4 encoding='UTF-8'>

```
https://jenkins.ceph.com/job/ceph-pull-requests/110979/consoleText

**Comment by mchangir** [comment]
@vshankar this is ready for review

**Comment by vshankar** [comment]
> @vshankar this is ready for review

Please add a test for the fix.

File: qa/tasks/cephfs/test_volumes.py [hunk 2]
```diff
@@ -7888,3 +7888,43 @@ def test_unparseable_metafile_on_legacy_to_v1_upgrade(self):
...
+    def test_empty_name_string_for_subvolumegroup_name_argument(self):
+        """
+        To test that an empty string is unacceptable for a subvolumegroup name
+        argument when creating a subvolume
+        """
+        with self.assertRaises(CommandFailedError):
+            self._fs_cmd("subvolume", "create", self.volname, "sv1", "--group_name")
+
+        with self.assertRaises(CommandFailedError):
+            self._fs_cmd("subvolume", "create", self.volname, "sv1", "--group_name", "''")
```
Comment (dparmar18) [inline qa/tasks/cephfs/test_volumes.py:7930]: Can't we add these test cases to their relevant classes? That way we can avoid creating a new class and the helper function that can be those classes too

Comment (mchangir) [reply qa/tasks/cephfs/test_volumes.py:7930]: * creating a new class doesn't impose extra load during job run, so it should be okay
* the new class has been placed in `test_volumes.py` since the tests are relevant to volumes
* the `_fs_cmd()` function is used just as an abbreviation to the longer length function call that it uses; I could use the original function call as well
* a separate class will also help with fine grained testing for this specific feature while starting a suite run

File: src/pybind/ceph_argparse.py [hunk 3]
```diff
@@ -327,7 +327,7 @@ class CephString(CephArgtype):
     """
     String; pretty generic.  goodchars is a RE char class of valid chars
     """
-    def __init__(self, goodchars=''):
+    def __init__(self, goodchars='', allowempty=True):
```
Comment (dparmar18) [inline src/pybind/ceph_argparse.py:330]: We want to allow empty strings by default? Maybe we can keep this false by default?

Comment (vshankar) [reply src/pybind/ceph_argparse.py:330]: The default behavior is to allow empty strings.

Comment (mchangir) [reply src/pybind/ceph_argparse.py:330]: I intend to keep the default behavior and not cause the tests grind to a halt due to changes to the default behavior.

**Comment by mchangir** [comment]
[Teuthology job](http://pulpito.front.sepia.ceph.com/mchangir-2023-02-22_13:27:59-fs:volumes-wip-mchangir-mds-handle-bad-arguments-during-subvolume-create-distro-default-smithi/): passed

@vshankar ping

**Comment by vshankar** [comment]
@ljflores This touches ceph_argparse python source, would you be the one to review that?

**Comment by vshankar** [comment]
jenkins test make check

**Comment by mchangir** [comment]
updated commit title and message

File: qa/tasks/cephfs/test_volumes.py [hunk 4]
```diff
@@ -7888,3 +7888,46 @@ def test_unparseable_metafile_on_legacy_to_v1_upgrade(self):
...
+    def tearDown(self):
+        # clean up
+        super().tearDown()
+
+    def test_empty_name_string_for_subvolumegroup_name(self):
+        """
+        To test that an empty string is unacceptable for a subvolumegroup name
+        """
+        with self.assertRaises(CommandFailedError):
+            self._fs_cmd("subvolumegroup", "create", self.volname, "''")
```
Comment (rishabh-d-dave) [inline qa/tasks/cephfs/test_volumes.py:7915]: nit: Wouldn't it be better if we check both -- empty argument as well empty quotes -- like the way we do in test method `test_empty_name_string_for_subvolumegroup_name_argument`?
```suggestion
        with self.assertRaises(CommandFailedError):
            self._fs_cmd("subvolumegroup", "create", self.volname, "''")
        with self.assertRaises(CommandFailedError):
            self._fs_cmd("subvolumegroup", "create", self.volname, "")            
```

File: qa/tasks/cephfs/test_volumes.py [hunk 5]
```diff
@@ -7888,3 +7888,46 @@ def test_unparseable_metafile_on_legacy_to_v1_upgrade(self):
...
+
+    def test_empty_name_string_for_subvolume_name(self):
+        """
+        To test that an empty string is unacceptable for a subvolume name
+        """
+        with self.assertRaises(CommandFailedError):
+            self._fs_cmd("subvolume", "create", "")
```
Comment (rishabh-d-dave) [inline qa/tasks/cephfs/test_volumes.py:7922]: nit: Same. Why not check both empty arguments as well empty quotes?
```suggestion
        with self.assertRaises(CommandFailedError):
            self._fs_cmd("subvolume", "create", "")
        with self.assertRaises(CommandFailedError):
            self._fs_cmd("subvolume", "create", "''")
```

**Review by rishabh-d-dave — APPROVED** [review]
Looks good to me.

File: qa/tasks/cephfs/test_volumes.py (hunk 4 above)
Comment (mchangir) [reply qa/tasks/cephfs/test_volumes.py:7918]: added the test case

File: qa/tasks/cephfs/test_volumes.py (hunk 5 above)
Comment (mchangir) [reply qa/tasks/cephfs/test_volumes.py:7925]: added the test case

**Comment by mchangir** [comment]
[Teuthology job](http://pulpito.front.sepia.ceph.com/mchangir-2023-02-24_01:28:14-fs:volumes-wip-mchangir-mds-handle-bad-arguments-during-subvolume-create-distro-default-smithi/): passed

**Comment by mchangir** [comment]
jenkins test make check

**Review by rishabh-d-dave — APPROVED** [review]

**Review by dparmar18 — APPROVED** [review]
LGTM

**Comment by vshankar** [comment]
@ljflores ping?

**Comment by vshankar** [comment]
@ljflores PTAL

**Comment by vshankar** [comment]
@mchangir please rebase.

**Comment by vshankar** [comment]
> @mchangir please rebase.

@mchangir ping.

**Comment by mchangir** [comment]
rebased

**Comment by vshankar** [comment]
https://pulpito.ceph.com/?branch=wip-vshankar-testing-20230421.044040

**Comment by vshankar** [comment]
jenkins test windows

**Comment by vshankar** [comment]
https://pulpito.ceph.com/?branch=wip-vshankar-testing-20230509.090957

**Comment by R-0-Y** [comment]
> https://pulpito.ceph.com/?branch=wip-vshankar-testing-20230509.090957

Hi @vshankar , do you have any news on this PR ? We're affected by the bug it fixes.
Thanks for your time on it.

**Comment by vshankar** [comment]
> > https://pulpito.ceph.com/?branch=wip-vshankar-testing-20230509.090957
> 
> Hi @vshankar , do you have any news on this PR ? We're affected by the bug it fixes. Thanks for your time on it.

fs suite run here: https://pulpito.ceph.com/vshankar-2023-06-21_05:44:47-fs-wip-vshankar-testing-20230620.131636-testing-default-smithi/

(requires verification before merging). Will update soon.

**Comment by vshankar** [comment]
Test passes: https://pulpito.ceph.com/vshankar-2023-06-21_05:44:47-fs-wip-vshankar-testing-20230620.131636-testing-default-smithi/7309526/

**Review by vshankar — APPROVED** [review]

**Comment by vshankar** [comment]
> @ljflores This touches ceph_argparse python source, would you be the one to review that?
> 
> EDIT: Changes LGTM, but wanted someone from the relevant team to have a look.

@ljflores ping? Could you please have a look.

**Comment by vshankar** [comment]
> Test passes: https://pulpito.ceph.com/vshankar-2023-06-21_05:44:47-fs-wip-vshankar-testing-20230620.131636-testing-default-smithi/7309526/

I'll take that back. @mchangir Please have a look at - https://pulpito.ceph.com/vshankar-2023-06-21_05:44:47-fs-wip-vshankar-testing-20230620.131636-testing-default-smithi/7309397/

Seems related to this change.

**Review by vshankar — CHANGES_REQUESTED** [review]
https://github.com/ceph/ceph/pull/50028#issuecomment-1604185108

**Comment by mchangir** [comment]
Fixed a bug introduced by commit bf83eaa4e755. The commit had introduced a spurious empty string as a command-line argument to `fs subvolume create` command if no `subvol_options` section was found in the job config.

[A sanity job run mostly passes with the changes.](https://pulpito.ceph.com/mchangir-2023-06-27_14:05:47-fs:workload-wip-mchangir-mds-handle-bad-arguments-during-subvolume-create-distro-default-smithi/)

**Comment by vshankar** [comment]
> Fixed a bug introduced by commit [bf83eaa](https://github.com/ceph/ceph/commit/bf83eaa4e75516a6937e4097b8708c48856a9473). The commit had introduced a spurious empty string as a command-line argument to `fs subvolume create` command if no `subvol_options` section was found in the job config.
> 
> [A sanity job run mostly passes with the changes.](https://pulpito.ceph.com/mchangir-2023-06-27_14:05:47-fs:workload-wip-mchangir-mds-handle-bad-arguments-during-subvolume-create-distro-default-smithi/)

Thx @mchangir. I think there is merit running this change through other subsystem suites.

**Comment by vshankar** [comment]
ping @ljflores

**Comment by mchangir** [comment]
rebased

**Comment by vshankar** [comment]
@mchangir

**Comment by vshankar** [comment]
ropening...
